data_model/src/query/cursor.rs

```diff
@@ -0,0 +1,98 @@
+//! Structures and traits related to server-side cursor.
+
+use core::num::{NonZeroU64, NonZeroUsize};
+
+use getset::Getters;
+use iroha_data_model_derive::model;
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode, Input};
+use serde::Serialize;
+
+pub use self::model::*;
+
+const QUERY_ID: &str = "query_id";
+const CURSOR: &str = "cursor";
+
+#[model]
+pub mod model {
+    use super::*;
+
+    /// Forward-only (a.k.a non-scrollable) cursor
+    #[derive(Debug, Clone, PartialEq, Eq, Default, Getters, Encode, Serialize, IntoSchema)]
+    #[getset(get = "pub")]
+    pub struct ForwardCursor {
+        /// Unique ID of query. When provided in a query the query will look up if there
+        /// is was already a query with a matching ID and resume returning result batches
+        pub query_id: Option<String>,
+        /// Pointer to the next element in the result set
+        pub cursor: Option<NonZeroU64>,
+    }
+}
+
+mod candidate {
+    use serde::{de::Error as _, Deserialize};
+
+    use super::*;
+
+    #[derive(Decode, Deserialize)]
+    struct ForwardCursorCandidate {
+        query_id: Option<String>,
+        cursor: Option<NonZeroU64>,
+    }
+
+    impl<'de> Deserialize<'de> for ForwardCursor {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: serde::Deserializer<'de>,
+        {
+            let candidate = ForwardCursorCandidate::deserialize(deserializer)?;
+
+            if let Some(query_id) = candidate.query_id {
+                Ok(ForwardCursor {
+                    query_id: Some(query_id),
+                    cursor: candidate.cursor,
+                })
+            } else if candidate.cursor.is_some() {
+                Err(D::Error::custom("Cursor missing query id"))
+            } else {
+                Ok(ForwardCursor::default())
```

Comment:
What's the meaning of default forward cusor? Shoulnd't it always be attached to some query?

Reply:
it's an empty/unused cursor. It has no query_id defined

Reply:
And why do we need this kind of unitialized cursor? Maybe the code that uses it should use `Option` instead?

Reply:
well honestly, dealing with query strings in warp is a pain. It was rejecting my other attempts, this was a thing that I found works

Reply:
Understandable